File: modules/editor/web/js/main-elements/life-lines/resource.js
/**
 * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
define(['d3'], function (d3) {

    var ResourceLifeline = {
        id: "Resource",
        title: "Pipe Line",
        icon: "images/tool-icons/lifeline.svg",
        class : "resource",
        shape: 'rect',
        editable: true,
        deletable: true,
        dragCursorOffset : { left: 50, top: 50 },
        createCloneCallback : function(view){
            function cloneCallBack() {
                var div = view.createContainerForDraggable();
                d3.xml("images/tool-icons/lifeline.svg").mimeType("image/svg+xml").get(function(error, xml) {
                    if (error) throw error;
                    var svg = xml.getElementsByTagName("svg")[0];
                    d3.select(svg).attr("width", "100px").attr("height", "100px");
                    d3.select(svg).attr("width", "100px");
                    div.node().appendChild(svg);
                });
                return div.node();
            }
            return cloneCallBack;
        },
        parameters: [
            {
                key: "title",
                value: "Resource"
            },
            {
                key: "path",
                value: ""
            },
            {
                key: "get",
                value: false
            },
            {
                key: "put",
                value: false
            },
            {
                key: "post",
                value: false
            }
        ],
        propertyPaneSchema: [
            {
                key: "title",
                text: "Title"
            },
            {
                key: "path",
                text: "Path"
            },
            {
                key: "get",
                checkbox: "GET"
            },
            {
                key: "put",
                checkbox: "PUT"
            },
            {
                key: "post",
                checkbox: "POST"
            }
        ],
        textModel : "undefined",
        utils: {
            getMyPropertyPaneSchema : function () {
                return MainElements.lifelines.ResourceLifeline.propertyPaneSchema;
            },
            getMyParameters: function (model) {
                return model.attributes.parameters;
            },
            saveMyProperties: function (model, inputs) {
                model.attributes.title = inputs.title.value;
                model.attributes.parameters = [
                    {
                        key: "title",
                        value: inputs.title.value
                    },
                    {
                        key: "path",
                        value: inputs.path.value
                    },
                    {
                        key: "get",
                        value: inputs.get.checked
                    },
                    {
                        key: "put",
                        value: inputs.put.checked
                    },
                    {
                        key: "post",
                        value: inputs.post.checked
                    }
                ];
            },
            canConnectTo: function () {
                return ['Worker', 'Source', 'ContainableProcessorElement', 'EndPoint'];
            },
            createMyModel: function (model) {
                //add the resource model
                var resourceCenterPoint = createPoint(380, 50);
                var resourceType = "Resource";
                var resourceLifeLineDef = MainElements.lifelines.ResourceLifeline;

                var resourceLifeline = createLifeLine(resourceType, resourceCenterPoint, resourceLifeLineDef.class,
                    resourceLifeLineDef.utils,
                    resourceLifeLineDef.parameters, resourceLifeLineDef.textModel,
                    resourceType, resourceLifeLineDef);
                var resourceLifeLineOptions = {
                    class: MainElements.lifelines.ResourceLifeline.class,
                    diagram: model
                };
                model.addElement(resourceLifeline, resourceLifeLineOptions);

                //increment resource counter
                var resourceCount = model.resourceLifeLineCounter();
                resourceCount++;
                model.resourceLifeLineCounter(resourceCount);
                return resourceLifeline;
            },
            createMyStartProcessorModel : function (resourceLifeline) {
                var position = new GeoCore.Models.Point({x: 0, y: 0});
                var processor = new SequenceD.Models.ProcessorFactory(Processors.actions.startAction.title,
                    position,
                    Processors.actions.startAction.type,
                    {
                        type: Processors.actions.startAction.type,
                        initMethod: Processors.actions.startAction.init
                    },
                    {
                        colour: Processors.actions.startAction.colour
                    },
                    Processors.actions.startAction.parameters,
                    Processors.actions.startAction.utils,
                    Processors.actions.startAction.textModel,
                    Processors.actions.startAction.width,
                    Processors.actions.startAction.height);

                resourceLifeline.addChild(processor);
                return processor;
            }
        }
    };

    return ResourceLifeline;
});

File: tests/observability-test-utils/src/main/java/org/ballerina/testobserve/listenerendpoint/Resource.java
/*
 * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.ballerina.testobserve.listenerendpoint;

import io.ballerina.runtime.api.types.ResourceFunctionType;
import io.ballerina.runtime.api.types.Type;
import io.ballerina.runtime.api.values.BObject;

import java.util.Objects;

import static io.ballerina.runtime.api.constants.TypeConstants.STRING_TNAME;
import static org.ballerina.testobserve.listenerendpoint.Constants.CALLER_TYPE_NAME;
import static org.ballerina.testobserve.listenerendpoint.Constants.TEST_OBSERVE_PACKAGE;

/**
 * Holds information related to a Ballerina service resource.
 */
public class Resource {
    private final BObject serviceObject;
    private final ResourceFunctionType resourceFunctionType;
    private final String basePath;

    public Resource(BObject serviceObject, ResourceFunctionType resourceFunctionType, String basePath) {
        this.serviceObject = serviceObject;
        this.resourceFunctionType = resourceFunctionType;
        this.basePath = basePath;

        // Validate resource
        int paramCount = this.getParamTypes().length;
        String resourcePath = this.getResourcePath();
        if (paramCount > 2) {
            throw new IllegalArgumentException("Invalid number of arguments in resource function "
                    + resourcePath + ". Expected a maximum of 2 arguments, but found " + paramCount);
        }
        if (paramCount >= 1) {
            Type paramOneType = this.getParamTypes()[0];
            if (!Objects.equals(paramOneType.getPackage(), TEST_OBSERVE_PACKAGE)
                    || !Objects.equals(paramOneType.getName(), CALLER_TYPE_NAME)) {
                throw new IllegalArgumentException("Invalid first parameter in the resource function "
                        + resourcePath + ". Expected parameter of type " + TEST_OBSERVE_PACKAGE + "/"
                        + CALLER_TYPE_NAME + ", but found " + paramOneType.getPackage() + "/"
                        + paramOneType.getName());
            }
        }
        if (paramCount >= 2) {
            Type paramTwoType = this.getParamTypes()[1];
            if (!Objects.equals(paramTwoType.getPackage().toString(), "null")
                    || !Objects.equals(paramTwoType.getName(), STRING_TNAME)) {
                throw new IllegalArgumentException("Invalid second parameter in the resource function "
                        + resourcePath + ". Expected parameter of type \"" + STRING_TNAME + "\", but found \""
                        + paramTwoType.getName() + "\"");
            }
        }
    }

    public BObject getServiceObject() {
        return this.serviceObject;
    }

    public String getResourceFunctionName() {
        return this.resourceFunctionType.getName();
    }

    public String getAccessor() {
        return this.resourceFunctionType.getAccessor();
    }

    public String getResourcePath() {
        return Utils.normalizeResourcePath(basePath)
                + Utils.normalizeResourcePath(String.join("/", this.resourceFunctionType.getResourcePath()));
    }

    public Type[] getParamTypes() {
        return this.resourceFunctionType.getParameterTypes();
    }

    public Type getReturnType() {
        return this.resourceFunctionType.getReturnType();
    }
}
